universal/include/userver/formats/universal/common_containers.hpp

```diff
@@ -0,0 +1,296 @@
+  };
+  template <typename Value>
+  inline constexpr std::enable_if_t<meta::kIsRange<Value>, std::string>
+  Check(const Value& value) const {
+    for(const auto& element : value) {
+      if(element < this->value) {
+        return this->Error(element);
+      }
+    }
+    return "";
+  }
+  inline constexpr auto Error(const T& value) const {
+    return std::format("{} < {}", value, this->value);
+  };
+};
+template <typename T>
+struct Default : public impl::EmptyCheck, public impl::Param<T> {
+  inline constexpr Default(const T& value) :
+      impl::Param<T>(value) {}
+};
+template <utils::ConstexprString Pattern>
+static const utils::regex kRegex(Pattern);
+
+struct Pattern : public impl::EmptyCheck, public impl::Param<const utils::regex*> {
+  constexpr inline Pattern(const utils::regex& regex) :
+      impl::Param<const utils::regex*>(&regex) {}
+  constexpr inline std::string Check(std::string_view str) const {
+    return utils::regex_match(str, *this->value) ? "" : this->Error(str);
+  }
+  constexpr inline std::string Error(std::string_view) const {
+    return "Error";
+  }
+};
+struct Additional : public impl::EmptyCheck, public impl::Param<bool> {
+  constexpr inline Additional(const bool& value) :
+      impl::Param<bool>(value) {}
+};
+template <>
+struct FieldConfig<int> {
+  std::optional<Max<int>> Maximum;
+  std::optional<Min<int>> Minimum;
+  template <typename MainClass, auto I, typename Value>
+  constexpr int Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return value[name].template As<int>();
+  };
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<int> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return parse::TryParse(value[name], parse::To<int>{});
+  };
+  constexpr auto Write(const int& value, std::string_view fieldName, const auto&, auto& builder) const {
+    builder[static_cast<std::string>(fieldName)] = value;
+  };
+  inline constexpr std::string_view Check(const int&) const {
+    return "";
+  }
+
+};
+template <>
+struct FieldConfig<std::optional<std::string>> {
+  std::optional<Pattern> Pattern;
+  std::optional<Default<std::string>> Default;
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::string> Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    if(!value.HasMember(name)) {
+      return std::nullopt;
+    };
+    return value[name].template As<std::string>();
+  };
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::optional<std::string>> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    auto response = parse::TryParse(value[name], parse::To<std::string>{});
+    if(response) {
+      return response;
+    }
+    if(this->Default) {
```

Comment:
default should only be used when the field is missing in the value (not even null)

If there is something, and that "something" fails to be parsed, then the function should just return failure

Reply:
Actually, `optional` can map to 2 things in OpenAPI:

* `!required`, meaning that it's allowed to be `IsMissing` (but not `null`), in which case `IsMissing() <-> nullopt`
* `nullable`, meaning that it's allowed to be `IsNull` (but not missing), in which case `IsNull() <-> nullopt`
* Both `!required` and `nullable`, in which case `IsMissing() -> nullopt`, `IsNull -> nullopt`
   - In this case there's a separate good question - to what `nullopt` should be serialized - missing or `null`. Let's pick "missing" for now

Reply:
So we need two extra config fields for `optional`:

- `.Required`, defaulting to `true`
- `.Nullable`, defaulting to `false`

In the case of `Required && !Nullable`, the whole `optional` is kind of meaningless, but let's allow it, I guess. It should throw when trying to Serialize `nullopt`